datadog/client.go

```diff
@@ -1,103 +1,176 @@
-	} else {
-		log.Printf("stats/datadog: connection opened to %s with a buffer size of %d B", config.Address, cap(conn.b))
+	if config.BufferSize == 0 {
+		config.BufferSize = DefaultBufferSize
 	}
 
-	return &Client{
-		conn: conn,
+	c := &Client{
+		serializer: serializer{},
+	}
+
+	conn, bufferSize, err := dial(config.Address, config.BufferSize)
+	if err != nil {
+		log.Printf("stats/datadog: %s", err)
 	}
+
+	c.conn, c.err = conn, err
+	c.buffer.BufferSize = bufferSize
+	c.buffer.Serializer = &c.serializer
+	return c
 }
 
-// Close satisfies the io.Closer interface.
-func (c *Client) Close() (err error) {
-	c.once.Do(func() {
-		if c.conn != nil {
-			err = c.conn.Close()
-		}
-	})
-	return
+// HandleMetric satisfies the stats.Handler interface.
+func (c *Client) HandleMeasures(time time.Time, measures ...stats.Measure) {
+	c.buffer.HandleMeasures(time, measures...)
 }
 
-// Flsuh satisfies the stats.Flusher interface.
+// Flush satisfies the stats.Flusher interface.
 func (c *Client) Flush() {
-	if c.conn != nil {
-		if err := c.conn.Flush(); err != nil {
-			log.Printf("stats/datadog: sending metrics to %s failed: %s", c.conn.RemoteAddr(), err)
-		}
+	c.buffer.Flush()
+}
+
+// Close flushes and closes the client, satisfies the io.Closer interface.
+func (c *Client) Close() error {
+	c.Flush()
+	c.close()
+	return c.err
+}
+
+type serializer struct {
+	conn net.Conn
+}
+
+func (*serializer) AppendMeasures(b []byte, _ time.Time, measures ...stats.Measure) []byte {
+	for _, m := range measures {
+		b = AppendMeasure(b, m)
 	}
+	return b
 }
 
-// HandleMetric satisfies the stats.Handler interface.
-func (c *Client) HandleMetric(m *stats.Metric) {
-	if c.conn != nil {
-		buf := bufferPool.Get().(*buffer)
-		buf.b = appendMetric(buf.b[:0], Metric{
-			Type:      metricType(m),
-			Namespace: m.Namespace,
-			Name:      m.Name,
-			Value:     m.Value,
-			Tags:      m.Tags,
-		})
-		if _, err := c.conn.Write(buf.b); err != nil {
-			log.Printf("stats/datadog: sending metric %s to %s failed: %s", m.Name, c.conn.RemoteAddr(), err)
+func (s *serializer) Write(b []byte) (int, error) {
+	if s.conn != nil {
+		return s.conn.Write(b)
+	}
+	return 0, io.ErrClosedPipe
+}
+
+func (s *serializer) close() {
+	if s.conn != nil {
+		s.conn.Close()
+	}
+}
+
+func dial(address string, sizehint int) (conn net.Conn, bufsize int, err error) {
+	var f *os.File
+
+	if conn, err = net.Dial("udp", address); err != nil {
+		return
+	}
+
+	if f, err = conn.(*net.UDPConn).File(); err != nil {
+		conn.Close()
+		return
+	}
+	defer f.Close()
+	fd := int(f.Fd())
+
+	// The kernel refuses to send UDP datagrams that are larger than the size of
+	// the size of the socket send buffer. To maximize the number of metrics
+	// sent in one batch we attempt to attempt to adjust the kernel buffer size
+	// to accept larger datagrams, or fallback to the default socket buffer size
+	// if it failed.
```

Comment:
Is this a bug in Go?  I'd expect the datagram to be split up and sent separately if this is the case.

I suppose since it's UDP there's no good way for them to decide where to split the message though.